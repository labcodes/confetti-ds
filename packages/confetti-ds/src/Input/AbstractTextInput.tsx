/* eslint-disable react/no-did-update-set-state */
import React, { SyntheticEvent } from "react";
import { isUndefined } from "lodash";
import Icon from "../Icon";

import { IconTypes, IconColorTypes } from "../constants";
import { usePrevious } from "../hooks";
import { BaseTextInputProps } from "./types";

export interface AbstractTextInputProps extends BaseTextInputProps {
  /** Passes AbstractInput's type to the HTML Input `type` attribute of the `<input>` element. */
  type?: string;
  /** Type of the icon to be rendered. Won't render an icon if not passed to the component. */
  icon?: IconTypes;
  /** Defines the color of the displayed icon. */
  iconColor?: IconColorTypes;
  /** Defines if the Input is valid. */
  isValid?: boolean;
  /** Custom error message displayed below the Input when the value is not valid. */
  customErrorMsg?: string;
  /** Callback action to be executed when the Input's Icon is clicked. */
  onIconClick?: (event?: SyntheticEvent) => any;
}

export default function AbstractTextInput({
  id,
  type = "text",
  icon,
  iconColor = "mineral-70" as const,
  isValid,
  customErrorMsg,
  onIconClick = () => {},
  ariaDisabled = false,
  disabled = false,
  required = false,
  onChange = () => {},
  defaultValue,
  value,
  prefix,
  suffix,
  label,
  helpMessage,
}: AbstractTextInputProps) {
  const [localValue, setLocalValue] = React.useState(
    value || defaultValue || ""
  );
  const [localIsValid, setLocalIsValid] = React.useState(
    !isUndefined(isValid) ? isValid : true
  );

  const inputRef = React.useRef<HTMLInputElement>();
  const previousLocalValue = usePrevious(localValue);

  let className = "";
  if (ariaDisabled || disabled) {
    className += " lab-input--disabled";
  } else if (!localIsValid) {
    className += " lab-input--invalid";
  }

  if (!isUndefined(defaultValue) && !isUndefined(value)) {
    // eslint-disable-next-line no-console
    console.warn(
      `You are setting both value and defaultValue for input ${id} at the same time. We always initialize value, if it is truthy. Make sure this is the behaviour you want.`
    );
  }

  React.useEffect(() => {
    if (inputRef.current) {
      /**
       * Has defaultValue but has no value and isValid === undefined
       */
      if (defaultValue && !value && isUndefined(isValid)) {
        const localIsValid = inputRef.current.validity.valid;
        setLocalIsValid(localIsValid);
      }

      /**
       * Has value but isValid === undefined
       */
      if (value && isUndefined(isValid)) {
        const localIsValid = inputRef.current.validity.valid;
        setLocalIsValid(localIsValid);
      }

      if (!localIsValid) {
        inputRef.current.setCustomValidity(customErrorMsg);
      }
    }

    return () => {
      setLocalIsValid(true);
      setLocalValue("");
    };
  }, []);

  React.useEffect(() => {
    if (!isUndefined(value)) setLocalValue(value);

    if (inputRef.current) {
      if (!isUndefined(isValid) && !isValid) {
        setLocalIsValid(isValid);
        inputRef.current.setCustomValidity(customErrorMsg);
      } else if (isValid) {
        setLocalIsValid(isValid);
        inputRef.current.setCustomValidity("");
      }
    }
  }, [isValid, value]);

  React.useEffect(() => {
    const shouldChangeValidity =
      !isUndefined(previousLocalValue) && previousLocalValue !== localValue;
    if (shouldChangeValidity) {
      setLocalIsValid(inputRef.current.validity.valid);
    }
  }, [localValue]);

  const requiredIcon = () =>
    required ? (
      <span className="lab-input__required-icon">
        <Icon type="star" color="white" />
      </span>
    ) : (
      ""
    );

  const prefixArea = () =>
    prefix ? <span className="lab-input__prefix">{prefix}</span> : "";

  const suffixArea = () =>
    suffix ? <div className="lab-input__suffix">{suffix}</div> : "";

  const handleOnChange = (event) => {
    const inputElement = event.target;
    const inputElementValue = inputElement.value;
    const inputElementIsValid = inputElement.validity.valid;

    // First we reset the custom validity
    inputElement.setCustomValidity("");

    if (!isUndefined(onChange)) {
      onChange(event);
    }

    // Then we set the state with the new value
    setLocalValue(inputElementValue);

    if (isUndefined(isValid) || (isValid && required)) {
      // Finally, if the user doesn't force the 'isValid', we use browser's validation from the input
      setLocalIsValid(inputElementIsValid);
    } else if (!isValid) {
      // We only set the customErrorMsg again if the input is forced invalid
      inputElement.setCustomValidity(customErrorMsg);
      setLocalIsValid(isValid);
    }
  };

  return (
    <React.Fragment>
      <div className={`lab-input ${className}`}>
        <input
          className={
            `lab-input__field ` +
            `${prefix ? `lab-input__field--prefixed ` : ``}` +
            `${suffix ? `lab-input__field--suffixed ` : ``}`
          }
          id={id}
          name={id}
          type={type}
          value={localValue}
          ref={inputRef}
          onChange={!ariaDisabled ? handleOnChange : () => {}}
          autoComplete="off"
          placeholder=" " // required for label placement
          disabled={(!ariaDisabled && disabled) || undefined}
          aria-disabled={ariaDisabled || undefined}
          required={required || undefined}
        />
        <div className="lab-input__borders" />
        {prefixArea()}
        {suffixArea()}
        <div className="lab-input__label-wrapper">
          {/* The following duplicated prefixArea is necessary to allow the label to be positioned correctly */}
          {prefixArea()}
          <label className="lab-input__label" htmlFor={id}>
            {label}
          </label>
        </div>
        {icon ? (
          <TrailingIcon
            icon={icon}
            iconColor={iconColor}
            onIconClick={onIconClick}
            disabled={(!ariaDisabled && disabled) || undefined}
            ariaDisabled={ariaDisabled || undefined}
          />
        ) : null}
        {requiredIcon()}
      </div>
      <TextInputMessage
        helpMessage={helpMessage}
        customErrorMsg={customErrorMsg}
        localValue={localValue}
        localIsValid={localIsValid}
      />
    </React.Fragment>
  );
}

// ----- Auxiliary components ----- //

interface TrailingIconProps {
  icon: IconTypes;
  iconColor?: IconColorTypes;
  onIconClick?: (event: SyntheticEvent) => any;
  disabled?: boolean;
  ariaDisabled?: boolean;
}

function TrailingIcon({
  iconColor = "mineral-70",
  onIconClick = () => {},
  disabled = false,
  ariaDisabled = false,
  icon,
}: TrailingIconProps) {
  return (
    <button
      type="button"
      className={`lab-input__icon${
        disabled ? ` lab-input__icon--disabled` : ``
      }`}
      onClick={!ariaDisabled ? onIconClick : () => {}}
      disabled={(!ariaDisabled && disabled) || undefined}
      aria-disabled={ariaDisabled || undefined}
    >
      <Icon type={icon} color={iconColor}/>
    </button>
  );
}

interface TextInputMessageProps {
  helpMessage: string;
  customErrorMsg: string;
  localValue: string;
  localIsValid: boolean;
}

function TextInputMessage({
  helpMessage,
  customErrorMsg,
  localIsValid,
  localValue,
}: TextInputMessageProps) {
  let message = null;
  if (helpMessage && localIsValid) {
    message = (
      <div className="lab-input__message lab-input__message--required">
        {helpMessage}
      </div>
    );
  }

  if (!localIsValid && helpMessage && !localValue) {
    message = (
      <div className="lab-input__message lab-input__message--error">
        {helpMessage}
      </div>
    );
  } else if (!localIsValid) {
    message = (
      <div className="lab-input__message lab-input__message--error">
        {customErrorMsg}
      </div>
    );
  }

  return message;
}
