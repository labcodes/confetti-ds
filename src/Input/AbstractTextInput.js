/* eslint-disable react/no-did-update-set-state */
import React from "react";
import PropTypes from "prop-types";
import { isUndefined } from "lodash";
import Icon from "../Icon";

export default class AbstractTextInput extends React.Component {
  static propTypes = {
    /** Passes AbstractInput's type to the HTML Input `type` attribute of the `<input>` element. */
    type: PropTypes.string,
    /** Text that will serve as unique identifier. It's also an important accessibility tool. */
    id: PropTypes.string.isRequired,
    /** The Input's text label. */
    label: PropTypes.string.isRequired,
    /** Disables the text input. Will be read by screen readers. When true, will override `disabled`. */
    ariaDisabled: PropTypes.bool,
    /** Disables the text input. Won't be read by screen readers. */
    disabled: PropTypes.bool,
    /** Defines a default value for the Input initialization. */
    defaultValue: PropTypes.string,
    /** Value that will be rendered inside the Input field. */
    value: PropTypes.string,
    /** Type of the icon to be rendered. Won't render an icon if not passed to the component. */
    icon: PropTypes.string,
    /** Defines the color of the displayed icon. */
    iconColor: PropTypes.string,
    /** Defines if the Input is required. */
    required: PropTypes.bool,
    /** Text that will be displayed as a help message below the input. */
    helpMessage: PropTypes.string,
    /** Text that will be displayed at the left portion of the Input. */
    prefix: PropTypes.string,
    /** Text that will be displayed at the right portion of the Input. */
    suffix: PropTypes.string,
    /** Defines if the Input is valid. */
    isValid: PropTypes.bool,
    /** Custom error message displayed below the Input when the value is not valid. */
    customErrorMsg: PropTypes.string,
    /** Callback action to be executed when the Input default value changes. */
    onChange: PropTypes.func,
    /** Callback action to be executed when the Input's Icon is clicked. */
    onIconClick: PropTypes.func,
  };

  static defaultProps = {
    type: "text",
    disabled: false,
    ariaDisabled: false,
    defaultValue: undefined,
    value: undefined,
    icon: undefined,
    iconColor: "mineral-70",
    required: false,
    helpMessage: undefined,
    prefix: undefined,
    suffix: undefined,
    isValid: undefined,
    customErrorMsg: undefined,
    onChange: () => {},
    onIconClick: () => {},
  };

  constructor(props) {
    super(props);
    this.inputRef = React.createRef();
    const { id, defaultValue, value, isValid } = props;
    if (!isUndefined(defaultValue) && !isUndefined(value)) {
      // eslint-disable-next-line no-console
      console.warn(
        `You are setting both value and defaultValue for input ${id} at the same time. We always initialize value, if it is truthy. Make sure this is the behaviour you want.`
      );
    }
    this.state = {
      localValue: value || defaultValue || "",
      localIsValid: !isUndefined(isValid) ? isValid : true,
    };
  }

  componentDidMount() {
    const { defaultValue, value, customErrorMsg } = this.props;
    const { isValid } = this.props;
    let { localIsValid } = this.state;

    if (defaultValue && !value && isUndefined(isValid)) {
      localIsValid = this.inputRef.current.validity.valid;
      this.setState(() => ({ localIsValid }));
    }

    if (value && isUndefined(isValid)) {
      localIsValid = this.inputRef.current.validity.valid;
      this.setState(() => ({ localIsValid }));
    }

    if (!localIsValid) {
      const inputElement = this.inputRef.current;
      inputElement.setCustomValidity(customErrorMsg);
    }
  }

  componentDidUpdate(prevProps) {
    const { value, isValid, customErrorMsg } = this.props;
    const inputElement = this.inputRef.current;

    if (isValid !== prevProps.isValid) {
      this.setState({ localIsValid: isValid });

      if (!isUndefined(isValid) && !isValid) {
        inputElement.setCustomValidity(customErrorMsg);
      } else if (isValid) {
        inputElement.setCustomValidity("");
      }
    }

    if (value !== prevProps.value) {
      this.setState({ localValue: value }, () => {
        if (isUndefined(isValid)) {
          this.setState({ localIsValid: inputElement.validity.valid });
        }
      });
    }
  }

  requiredIcon = () => {
    const { required } = this.props;
    return required ? (
      <span className="lab-input__required-icon">
        <Icon type="star" color="white" />
      </span>
    ) : (
      ""
    );
  };

  prefixArea = () => {
    const { prefix } = this.props;
    return prefix ? <span className="lab-input__prefix">{prefix}</span> : "";
  };

  suffixArea = () => {
    const { suffix } = this.props;
    return suffix ? <div className="lab-input__suffix">{suffix}</div> : "";
  };

  handleOnChange = (event) => {
    const { onChange, isValid, customErrorMsg, required } = this.props;
    const inputElement = event.target;
    const inputElementValue = inputElement.value;
    const inputElementIsValid = inputElement.validity.valid;

    // First we reset the custom validity
    inputElement.setCustomValidity("");

    if (!isUndefined(onChange)) {
      onChange(event);
    }

    // Then we set the state with the new value
    this.setState({ localValue: inputElementValue }, () => {
      if (isUndefined(isValid) || (isValid && required)) {
        // Finally, if the user doesn't force the 'isValid', we use browser's validation from the input
        this.setState({ localIsValid: inputElementIsValid });
      } else if (!isValid) {
        // We only set the customErrorMsg again if the input is forced invalid
        inputElement.setCustomValidity(customErrorMsg);
        this.setState({ localIsValid: isValid });
      }
    });
  };

  render() {
    const {
      type,
      id,
      label,
      disabled,
      ariaDisabled,
      icon,
      iconColor,
      required,
      helpMessage,
      prefix,
      suffix,
      customErrorMsg,
      onIconClick,
    } = this.props;

    let className = "";

    const { localValue, localIsValid } = this.state;
    if (ariaDisabled || disabled) {
      className += " lab-input--disabled";
    } else if (!localIsValid) {
      className += " lab-input--invalid";
    }

    return (
      <React.Fragment>
        <div className={`lab-input ${className}`}>
          <input
            className={
              `lab-input__field ` +
              `${prefix ? `lab-input__field--prefixed ` : ``}` +
              `${suffix ? `lab-input__field--suffixed ` : ``}`
            }
            id={id}
            name={id}
            type={type}
            value={localValue}
            ref={this.inputRef}
            onChange={!ariaDisabled ? this.handleOnChange : () => {}}
            autoComplete="off"
            placeholder=" " // required for label placement
            disabled={(!ariaDisabled && disabled) || undefined}
            aria-disabled={ariaDisabled || undefined}
            required={required || undefined}
          />
          <div className="lab-input__borders" />
          {this.prefixArea()}
          {this.suffixArea()}
          <div className="lab-input__label-wrapper">
            {/* The following duplicated prefixArea is necessary to allow the label to be positioned correctly */}
            {this.prefixArea()}
            <label className="lab-input__label" htmlFor={id}>
              {label}
            </label>
          </div>
          {icon ? (
            <TrailingIcon
              icon={icon}
              iconColor={iconColor}
              onIconClick={onIconClick}
              disabled={(!ariaDisabled && disabled) || undefined}
              ariaDisabled={ariaDisabled || undefined}
            />
          ) : null}
          {this.requiredIcon()}
        </div>
        <TextInputMessage
          helpMessage={helpMessage}
          customErrorMsg={customErrorMsg}
          localValue={localValue}
          localIsValid={localIsValid}
        />
      </React.Fragment>
    );
  }
}

// ----- Auxiliary components ----- //

function TrailingIcon(props) {
  const { icon, iconColor, onIconClick, disabled, ariaDisabled } = props;
  return (
    <button
      type="button"
      className={`lab-input__icon${
        disabled ? ` lab-input__icon--disabled` : ``
      }`}
      onClick={!ariaDisabled ? onIconClick : () => {}}
      disabled={(!ariaDisabled && disabled) || undefined}
      aria-disabled={ariaDisabled || undefined}
    >
      <Icon type={icon} color={iconColor} />
    </button>
  );
}

TrailingIcon.propTypes = {
  icon: PropTypes.string.isRequired,
  iconColor: PropTypes.string,
  onIconClick: PropTypes.func,
  disabled: PropTypes.bool,
  ariaDisabled: PropTypes.bool,
};

TrailingIcon.defaultProps = {
  iconColor: "mineral-70",
  onIconClick: () => {},
  disabled: false,
  ariaDisabled: false,
};

function TextInputMessage(props) {
  const { helpMessage, customErrorMsg, localIsValid, localValue } = props;
  let message = null;
  if (helpMessage && localIsValid) {
    message = (
      <div className="lab-input__message lab-input__message--required">
        {helpMessage}
      </div>
    );
  }

  if (!localIsValid && helpMessage && !localValue) {
    message = (
      <div className="lab-input__message lab-input__message--error">
        {helpMessage}
      </div>
    );
  } else if (!localIsValid) {
    message = (
      <div className="lab-input__message lab-input__message--error">
        {customErrorMsg}
      </div>
    );
  }

  return message;
}

TextInputMessage.propTypes = {
  helpMessage: PropTypes.string,
  customErrorMsg: PropTypes.string,
  localValue: PropTypes.string,
  localIsValid: PropTypes.bool,
};

TextInputMessage.defaultProps = {
  helpMessage: undefined,
  customErrorMsg: undefined,
  localValue: undefined,
  localIsValid: undefined,
};
