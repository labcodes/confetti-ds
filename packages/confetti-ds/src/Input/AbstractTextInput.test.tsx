/* eslint-disable no-console */
import React from "react";
import { mount } from "enzyme";
import renderer from "react-test-renderer";

import AbstractTextInput from "./AbstractTextInput";

describe("AbstractTextInput", () => {
  const originalWarn = console.warn;
  afterEach(() => {
    console.warn = originalWarn;
  });

  it("renders with base props", async () => {
    const renderedComponent = renderer
      .create(<AbstractTextInput id="testInput" label="Test Input" />)
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });

  it("renders as expected when passing disabled as true", async () => {
    const mountedComponent = mount(
      <AbstractTextInput
        id="testInput"
        label="Test Input"
        defaultValue="default"
        disabled
      />
    );
    expect(mountedComponent.find("input[disabled]")).toHaveLength(1);
  });

  it("raises console.warn and sets state with value when passing value and defaultValue by props at the same time", async () => {
    console.warn = jest.fn();

    const mountedComponent = mount(
      <AbstractTextInput
        id="testInput"
        label="Test Input"
        defaultValue="default value"
        value="test value"
      />
    );

    expect(console.warn).toBeCalled();
    const inputElement = mountedComponent.find("input");
    expect(inputElement.render().attr("value")).toBe("test value");
  });

  it("sets state with value if it is passed by props", async () => {
    const mountedComponent = mount(
      <AbstractTextInput id="testInput" label="Test Input" value="test value" />
    );

    const inputElement = mountedComponent.find("input");
    expect(inputElement.render().attr("value")).toBe("test value");
  });

  it("sets state with defaultValue if it is passed by props and value is not passed by props", async () => {
    const mountedComponent = mount(
      <AbstractTextInput
        id="testInput"
        label="Test Input"
        defaultValue="default value"
      />
    );
    const inputElement = mountedComponent.find("input");
    expect(inputElement.render().attr("value")).toBe("default value");
  });

  it("sets state with isValid if it is passed by props", async () => {
    const mountedComponent = mount(
      <AbstractTextInput
        id="testInput"
        label="Test Input"
        defaultValue="default value"
        isValid={false}
      />
    );
    expect(mountedComponent.find(".lab-input--invalid")).toHaveLength(1);
  });

  it("sets localIsValid to false if defaultValue is invalid", async () => {
    const mountedComponent = mount(
      <AbstractTextInput
        id="testInput"
        label="Test Input"
        type="email"
        defaultValue="default value"
      />
    );
    expect(mountedComponent.find(".lab-input--invalid")).toHaveLength(1);
  });

  it("sets localIsValid to false if value is invalid", async () => {
    const mountedComponent = mount(
      <AbstractTextInput
        id="testInput"
        label="Test Input"
        type="email"
        value="test value"
      />
    );
    expect(mountedComponent.find(".lab-input--invalid")).toHaveLength(1);
  });

  it("sets localIsValid to true if value is valid and change it if value validity changes", async () => {
    const mountedComponent = mount(
      <AbstractTextInput
        id="testInput"
        label="Test Input"
        type="email"
        value="testvalue@g.com"
      />
    );
    expect(mountedComponent.find(".lab-input--invalid")).toHaveLength(0);

    mountedComponent.setProps({ value: "value changed" });
    mountedComponent.update();

    expect(mountedComponent.find(".lab-input--invalid")).toHaveLength(1);

    mountedComponent.setProps({ value: "valid@value.com" });
    mountedComponent.update();

    expect(mountedComponent.find(".lab-input--invalid")).toHaveLength(0);
  });

  it("shows help text", async () => {
    const mountedComponent = mount(
      <AbstractTextInput
        id="testInput"
        label="Test Input"
        type="email"
        value="testvalue@g.com"
        helpMessage="help message"
        required
      />
    );

    expect(mountedComponent.find(".lab-input__message--required").text()).toBe(
      "help message"
    );

    mountedComponent.setProps({ value: "" });
    mountedComponent.update();

    expect(mountedComponent.find(".lab-input__message--error").text()).toBe(
      "help message"
    );
  });

  it("shows custom error message when input is invalid", async () => {
    const mountedComponent = mount(
      <AbstractTextInput
        id="testInput"
        label="Test Input"
        type="email"
        value="testvalue@g.com"
        customErrorMsg="error message"
      />
    );

    expect(mountedComponent.find(".lab-input__message--error")).toHaveLength(0);

    mountedComponent.setProps({ value: "value changed" });
    mountedComponent.update();

    expect(mountedComponent.find(".lab-input__message--error").text()).toBe(
      "error message"
    );
  });

  it("changes localValue state when input changes", async () => {
    const mockOnChange = jest.fn();
    const mountedComponent = mount(
      <AbstractTextInput
        id="testInput"
        label="Test Input"
        value="truthy value"
        onChange={mockOnChange}
      />
    );

    const inputElement = mountedComponent.find("input");
    expect(inputElement.render().attr("value")).toBe("truthy value");
    expect(mockOnChange).not.toBeCalled();
    inputElement.simulate("change", {
      target: {
        value: "My new value",
        validity: { valid: true },
        setCustomValidity: jest.fn,
      },
    });
    expect(mockOnChange).toBeCalled();
    expect(inputElement.render().attr("value")).toBe("My new value");
  });

  it("sets input as invalid if it is required and is not filled, even if isValid is passed by props as true", async () => {
    const mountedComponent = mount(
      <AbstractTextInput
        id="testInput"
        label="Test Input"
        isValid
        required
      />
    );

    expect(mountedComponent.find(".lab-input--invalid")).toHaveLength(0);

    mountedComponent
      .find("input")
      .at(0)
      .simulate("change", {
        target: {
          value: "My new value",
          validity: { valid: true },
          setCustomValidity: jest.fn(),
        },
      });

    expect(mountedComponent.find(".lab-input--invalid")).toHaveLength(0);

    mountedComponent
      .find("input")
      .at(0)
      .simulate("change", {
        target: {
          value: "",
          validity: { valid: false },
          setCustomValidity: jest.fn(),
        },
      });

    expect(mountedComponent.find(".lab-input--invalid")).toHaveLength(1);
  });

  it("renders prefix when it is passed by props", async () => {
    const renderedComponent = renderer
      .create(
        <AbstractTextInput id="testInput" label="Test Input" prefix="R$" />
      )
      .toJSON();

    expect(renderedComponent).toMatchSnapshot();

    const mountedComponent = mount(
      <AbstractTextInput id="testInput" label="Test Input" prefix="R$" />
    );

    expect(mountedComponent.find("span.lab-input__prefix")).toHaveLength(2);
  });

  it("renders suffix when it is passed by props", async () => {
    const renderedComponent = renderer
      .create(
        <AbstractTextInput id="testInput" label="Test Input" suffix=".com" />
      )
      .toJSON();

    expect(renderedComponent).toMatchSnapshot();

    const mountedComponent = mount(
      <AbstractTextInput id="testInput" label="Test Input" suffix=".com" />
    );

    expect(mountedComponent.find("div.lab-input__suffix")).toHaveLength(1);
  });

  it("renders icon when it is passed by props", async () => {
    const renderedComponent = renderer
      .create(
        <AbstractTextInput id="testInput" label="Test Input" icon="star" />
      )
      .toJSON();

    expect(renderedComponent).toMatchSnapshot();

    const mountedComponent = mount(
      <AbstractTextInput id="testInput" label="Test Input" icon="star" />
    );

    expect(mountedComponent.find("span.lab-icon")).toHaveLength(1);
  });

  it("renders required icon when it is passed by props", async () => {
    const renderedComponent = renderer
      .create(
        <AbstractTextInput id="testInput" label="Test Input" icon="star" />
      )
      .toJSON();

    expect(renderedComponent).toMatchSnapshot();

    const mountedComponent = mount(
      <AbstractTextInput id="testInput" label="Test Input" required />
    );

    expect(mountedComponent.find("span.lab-input__required-icon")).toHaveLength(
      1
    );
  });

  it("doesn't trigger onChange if ariaDisabled", async () => {
    const mockOnChange = jest.fn();
    const mountedComponent = mount(
      <AbstractTextInput
        ariaDisabled
        id="testInput"
        label="Test Input"
        defaultValue="default"
        onChange={mockOnChange}
      />
    );

    expect(mountedComponent.find("input").render().attr("value")).toBe(
      "default"
    );
    expect(mockOnChange).not.toBeCalled();

    mountedComponent.find("input").at(0).simulate("change", { test: "event" });

    expect(mountedComponent.find("input").render().attr("value")).toBe(
      "default"
    );
    expect(mockOnChange).not.toBeCalled();
  });
});
