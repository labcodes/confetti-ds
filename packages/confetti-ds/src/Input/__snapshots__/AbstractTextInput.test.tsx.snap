// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AbstractTextInput renders icon when it is passed by props 1`] = `
<div
  className="lab-input "
>
  <input
    autoComplete="off"
    className="lab-input__field "
    id="testInput"
    name="testInput"
    onChange={[Function]}
    placeholder=" "
    type="text"
    value=""
  />
  <div
    className="lab-input__borders"
  />
  
  
  <div
    className="lab-input__label-wrapper"
  >
    
    <label
      className="lab-input__label"
      htmlFor="testInput"
    >
      Test Input
    </label>
  </div>
  <button
    className="lab-input__icon"
    onClick={[Function]}
    type="button"
  >
    <span
      className="lab-icon lab-icon--star"
    >
      <svg
        aria-hidden={true}
        className="___12fm75w_v8ls9a0 f1w7gpdv fez10in fg4l7m0"
        fill="currentColor"
        height={20}
        viewBox="0 0 20 20"
        width={20}
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10 2a8 8 0 1 1 0 16 8 8 0 0 1 0-16Zm0 1a7 7 0 1 0 0 14 7 7 0 0 0 0-14Zm0 9.5a.75.75 0 1 1 0 1.5.75.75 0 0 1 0-1.5ZM10 6a.5.5 0 0 1 .5.41V11a.5.5 0 0 1-1 .09V6.5c0-.28.22-.5.5-.5Z"
          fill="currentColor"
        />
      </svg>
    </span>
  </button>
  
</div>
`;

exports[`AbstractTextInput renders prefix when it is passed by props 1`] = `
<div
  className="lab-input "
>
  <input
    autoComplete="off"
    className="lab-input__field lab-input__field--prefixed "
    id="testInput"
    name="testInput"
    onChange={[Function]}
    placeholder=" "
    type="text"
    value=""
  />
  <div
    className="lab-input__borders"
  />
  <span
    className="lab-input__prefix"
  >
    R$
  </span>
  
  <div
    className="lab-input__label-wrapper"
  >
    <span
      className="lab-input__prefix"
    >
      R$
    </span>
    <label
      className="lab-input__label"
      htmlFor="testInput"
    >
      Test Input
    </label>
  </div>
  
</div>
`;

exports[`AbstractTextInput renders required icon when it is passed by props 1`] = `
<div
  className="lab-input "
>
  <input
    autoComplete="off"
    className="lab-input__field "
    id="testInput"
    name="testInput"
    onChange={[Function]}
    placeholder=" "
    type="text"
    value=""
  />
  <div
    className="lab-input__borders"
  />
  
  
  <div
    className="lab-input__label-wrapper"
  >
    
    <label
      className="lab-input__label"
      htmlFor="testInput"
    >
      Test Input
    </label>
  </div>
  <button
    className="lab-input__icon"
    onClick={[Function]}
    type="button"
  >
    <span
      className="lab-icon lab-icon--star"
    >
      <svg
        aria-hidden={true}
        className="___12fm75w_v8ls9a0 f1w7gpdv fez10in fg4l7m0"
        fill="currentColor"
        height={20}
        viewBox="0 0 20 20"
        width={20}
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10 2a8 8 0 1 1 0 16 8 8 0 0 1 0-16Zm0 1a7 7 0 1 0 0 14 7 7 0 0 0 0-14Zm0 9.5a.75.75 0 1 1 0 1.5.75.75 0 0 1 0-1.5ZM10 6a.5.5 0 0 1 .5.41V11a.5.5 0 0 1-1 .09V6.5c0-.28.22-.5.5-.5Z"
          fill="currentColor"
        />
      </svg>
    </span>
  </button>
  
</div>
`;

exports[`AbstractTextInput renders suffix when it is passed by props 1`] = `
<div
  className="lab-input "
>
  <input
    autoComplete="off"
    className="lab-input__field lab-input__field--suffixed "
    id="testInput"
    name="testInput"
    onChange={[Function]}
    placeholder=" "
    type="text"
    value=""
  />
  <div
    className="lab-input__borders"
  />
  
  <div
    className="lab-input__suffix"
  >
    .com
  </div>
  <div
    className="lab-input__label-wrapper"
  >
    
    <label
      className="lab-input__label"
      htmlFor="testInput"
    >
      Test Input
    </label>
  </div>
  
</div>
`;

exports[`AbstractTextInput renders with base props 1`] = `
<div
  className="lab-input "
>
  <input
    autoComplete="off"
    className="lab-input__field "
    id="testInput"
    name="testInput"
    onChange={[Function]}
    placeholder=" "
    type="text"
    value=""
  />
  <div
    className="lab-input__borders"
  />
  
  
  <div
    className="lab-input__label-wrapper"
  >
    
    <label
      className="lab-input__label"
      htmlFor="testInput"
    >
      Test Input
    </label>
  </div>
  
</div>
`;
